### PR TITLE
bsp: New BSP for M5Stack Core (BSP-594)

### PR DESCRIPTION
**Pull request by Ocean-lhy**
# ESP-BSP Pull Request checklist

> Note: For new BSPs create a PR with this [link](https://github.com/espressif/esp-bsp/compare/main...my-branch?quick_pull=1&template=pr_template_bsp.md).
- [x] Version of modified component bumped
- [ ] CI passing

# Change description
- bsp: New BSP for M5Core
- doc(m5stack_core_2): added configuration methods for 1.0 and 1.1

### CODE REVIEW

File: bsp/m5stack_core/idf_component.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+version: "2.7.0"
```
Comment (espzav) [inline bsp/m5stack_core/idf_component.yml:1]:
```suggestion
version: "1.0.0"
```
This is version of the BSP

File: bsp/m5stack_core/idf_component.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+targets:
+  - esp32
+
+tags:
+  - bsp
+
+dependencies:
+  idf: ">=5.0"
+  esp_lcd_ili9341: "^1"
+  # esp_lcd_touch_ft5x06: "^1"
```
Comment (espzav) [inline bsp/m5stack_core/idf_component.yml:14]: There isn't touch, remove it.

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+ * Return value:
+ * - esp_err_t: Error code indicating the result of the operation.
+ *
+ **************************************************************************************************/
+typedef enum {
+    BSP_FEATURE_LCD,
+    BSP_FEATURE_TOUCH,
+    BSP_FEATURE_SD,
+    BSP_FEATURE_SPEAKER,
+    BSP_FEATURE_BATTERY
```
Comment (espzav) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:358]: Remove features, which are enabled in default and cannot be enabled/disabled by SW.

File: bsp/m5stack_core/m5stack_core.c [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+static bool spi_initialized = false;
+
+esp_err_t bsp_i2c_init(void)
+{
+    /* I2C was initialized before */
+    if (i2c_initialized) {
+        return ESP_OK;
+    }
+
+    const i2c_config_t i2c_conf = {.mode             = I2C_MODE_MASTER,
```
Comment (espzav) [inline bsp/m5stack_core/m5stack_core.c:47]: Please use new I2C driver like here: https://github.com/espressif/esp-bsp/blob/9ba025e1361f3252dc2be8c1aafcf69746544fd1/bsp/m5stack_core_s3/m5stack_core_s3.c#L68

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+/* SD card */
+#define BSP_SD_MOSI           (GPIO_NUM_23)
+#define BSP_SD_MISO           (GPIO_NUM_19)
+#define BSP_SD_SCK            (GPIO_NUM_18)
+#define BSP_SD_CS             (GPIO_NUM_4)
+
+/* BUTTON */
+#define BSP_BUTTON_A          (GPIO_NUM_39)
+#define BSP_BUTTON_B          (GPIO_NUM_38)
+#define BSP_BUTTON_C          (GPIO_NUM_37)
```
Comment (espzav) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:76]: What about set buttons same like here: https://github.com/espressif/esp-bsp/blob/9ba025e1361f3252dc2be8c1aafcf69746544fd1/bsp/esp-box-lite/esp-box-lite.c#L375
Like a navigation buttons for LVGL in default.

**Review by espzav — COMMENTED** [review]
I make only fast review for now. I left some comments.
Would it work with my old M5Stack Core (2018.3)?

**Comment by Ocean-lhy** [comment]
> I make only fast review for now. I left some comments. Would it work with my old M5Stack Core (2018.3)?

Thanks for your review. I've updated the code based on your suggestions. PTAL and resolve the conversation if the issue has been resolved.

By the way, I think the bsp will work on your old M5Stack Core, although I don't have the original M5Stack Core on hand. I checked for version changes and there were no changes that would cause compatibility issues. Perhaps you can test it with a simple lvgl demo.

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+/**
+ * @file
+ * @brief ESP BSP: M5Stack Core
+ */
+
+#pragma once
+
+#include "sdkconfig.h"
+#include "driver/gpio.h"
+#include "driver/i2c.h"
```
Comment (tore-espressif) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:16]: Can we remove this deprecated header? (The new header is `driver/i2c_master.h`)

```suggestion
```

File: bsp/m5stack_core/m5stack_core.c [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+    BSP_ERROR_CHECK_RETURN_ERR(i2c_master_bus_add_device(i2c_handle, &ip5306_config, &ip5306_h));
+
+    i2c_initialized = true;
+
+    return ESP_OK;
+}
+
+esp_err_t bsp_i2c_deinit(void)
+{
+    BSP_ERROR_CHECK_RETURN_ERR(i2c_del_master_bus(i2c_handle));
```
Comment (tore-espressif) [inline bsp/m5stack_core/m5stack_core.c:78]: Can we also delete the `ip5306_h` device here? So we don't create a memory leak

File: bsp/m5stack_core/m5stack_core.c [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+
+    return ESP_OK;
+}
+
+esp_codec_dev_handle_t bsp_audio_codec_speaker_init(void)
+{
+    // Simple GPIO initialization
+    BSP_ERROR_CHECK_RETURN_NULL(bsp_audio_init(NULL));
+    return NULL;  // No longer using codec
+}
```
Comment (tore-espressif) [inline bsp/m5stack_core/m5stack_core.c:209]: * How can the user play something through the speaker, when we return NULL here?
* Can we remove the `esp_codec_dev` dependency if we do not use it?

**Review by tore-espressif — COMMENTED** [review]
Thank you for much for this BSP!

I had a brief look.
I found old M5Stack Core in my drawer, I'll test it soon

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+ *      - ESP_ERR_INVALID_ARG   NULL pointer or invalid configuration
+ *      - ESP_ERR_NOT_FOUND     No available I2S channel found
+ *      - ESP_ERR_NO_MEM        No memory for storing the channel information
+ *      - ESP_ERR_INVALID_STATE This channel has not initialized or already started
+ */
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 0, 0)
+esp_err_t bsp_audio_init(const i2s_config_t *i2s_config);
+#else
+esp_err_t bsp_audio_init(const i2s_std_config_t *i2s_config);
+#endif
```
Comment (tore-espressif) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:117]: Do we use I2S anywhere in the BSP?

Comment (tore-espressif) [reply bsp/m5stack_core/include/bsp/m5stack_core.h:117]: I see that the speaker is connected to ESP32 without any DAC...

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+#define BSP_CAPS_TOUCH          0
+#define BSP_CAPS_BUTTONS        1
+#define BSP_CAPS_AUDIO          1
+#define BSP_CAPS_AUDIO_SPEAKER  1
+#define BSP_CAPS_AUDIO_MIC      0
+#define BSP_CAPS_SDCARD         1
+#define BSP_CAPS_IMU            0
+
+/**************************************************************************************************
+ *  M5Stack-Core-2 pinout
```
Comment (tore-espressif) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:45]:
```suggestion
 *  M5Stack-Core pinout
```

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+#define BSP_I2C_SDA           (GPIO_NUM_21)
+/* Audio */
+#define BSP_I2S_SCLK          (GPIO_NUM_12)
+#define BSP_I2S_MCLK          (GPIO_NUM_0)
+#define BSP_I2S_LCLK          (GPIO_NUM_13)
+#define BSP_I2S_DOUT          (GPIO_NUM_15)
+#define BSP_I2S_DSIN          (GPIO_NUM_34)
+#define BSP_SPEAKER_IO        (GPIO_NUM_25)
+#define BSP_POWER_AMP_IO      (GPIO_NUM_NC)
+#define BSP_MUTE_STATUS       (GPIO_NUM_NC)
```
Comment (tore-espressif) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:58]: Can be removed?

**Comment by Ocean-lhy** [comment]
I submitted the code according to feedback to see if the problem was improved?

File: bsp/m5stack_core/include/bsp/m5stack_core.h (hunk 11 above)
Comment (Ocean-lhy) [reply bsp/m5stack_core/include/bsp/m5stack_core.h]: I removed it, would it be better to leave the pins of the I2S as defined interfaces?

File: bsp/m5stack_core/m5stack_core.c [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        .reset_gpio_num = BSP_LCD_RST,  // Shared with Touch reset
+        .color_space    = BSP_LCD_COLOR_SPACE,
+        .bits_per_pixel = BSP_LCD_BITS_PER_PIXEL,
+    };
+
+    ESP_GOTO_ON_ERROR(esp_lcd_new_panel_ili9341(*ret_io, &panel_config, ret_panel), err, TAG, "New panel failed");
+
+    esp_lcd_panel_reset(*ret_panel);
+    esp_lcd_panel_init(*ret_panel);
+    esp_lcd_panel_invert_color(*ret_panel, true);
```
Comment (tore-espressif) [inline bsp/m5stack_core/m5stack_core.c:290]: I tried this BSP with `examples/display` example and I had to change this to `false` to get correct colors.

Does it work on your device? Is it possible that I have some older version of M5Stack Core that does not need color inversion?
```suggestion
    esp_lcd_panel_invert_color(*ret_panel, false);
```

Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c:290]: image
I think it's because of the screen change?

Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c:290]: We should do something here to avoid that. Do you have any good ideas?

Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c:290]: I tried to change it to false and it turned into a reverse color in version 2.7😂, which I think is the difference between TN screens and IPS screens
image
image

Comment (tore-espressif) [reply bsp/m5stack_core/m5stack_core.c:290]: So it works for you correctly with `esp_lcd_panel_invert_color(*ret_panel, true);`, right?

I checked you code from here https://github.com/m5stack/M5Unified/blob/master/src/M5Unified.cpp and it does not differentiate between Core versions...

Ideally, we would identify the board version from software and the set the color inversion accordingly. If this is not possible, please add a comment there that users of older M5Stack Core must set the color inversion to `false`

Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c:284]: I added the inverse color related configuration in the new commit, and added the description in `Kconfig` and `README`

**Comment by tore-espressif** [comment]
@Ocean-lhy The code changes LGTM!

To fix your CI issues, please install pre-commit with `pip install pre-commit && pre-commit install`
you can then run it on you last eg 4 commits with `pre-commit run --from-ref HEAD~4 --to-ref HEAD`


Could you also please add this commit ? https://github.com/espressif/esp-bsp/commit/3da9487a6c44b9a8515b79477c57f5746e6afb19
It enables `examples/display` example for M5Stack Core board. It is then build in CI to gurantee error free compilation

**Comment by Ocean-lhy** [comment]
@tore-espressif  I really appreciate your guidance. I have solved the CI problem according to your instructions.

Regarding commit 3da9487, I have pulled it to make sure the CI is error-free.

File: bsp/m5stack_core/idf_component.yml [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+url: https://github.com/espressif/esp-bsp/tree/master/bsp/m5stack_core
+
+targets:
+  - esp32
+
+tags:
+  - bsp
+
+dependencies:
+  idf: ">=5.0"
```
Comment (tore-espressif) [inline bsp/m5stack_core/idf_component.yml:12]: @Ocean-lhy I'm sorry, I missed this one. You are using new version of the I2C driver that was introduced in v5.2

Please also add 'm5stack_core' to this list https://github.com/espressif/esp-bsp/blob/master/.build-test-rules.yml#L6

```suggestion
  idf: ">=5.2"
```

**Review by tore-espressif — APPROVED** [review]

File: bsp/m5stack_core/m5stack_core.c [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+    /* Initialize SPI */
+    ESP_RETURN_ON_ERROR(bsp_spi_init(config->max_transfer_sz), TAG, "");
+
+    ESP_LOGI(TAG, "Install panel IO");
+    const esp_lcd_panel_io_spi_config_t io_config = {
+        .dc_gpio_num       = BSP_LCD_DC,
+        .cs_gpio_num       = BSP_LCD_CS,
+        .pclk_hz           = BSP_LCD_PIXEL_CLOCK_HZ,
+        .lcd_cmd_bits      = LCD_CMD_BITS,
+        .lcd_param_bits    = LCD_PARAM_BITS,
```
Comment (espzav) [inline bsp/m5stack_core/m5stack_core.c:265]: I tried to compile it and tried on my HW and I am missing these defines:
```
#define LCD_CMD_BITS           8
#define LCD_PARAM_BITS         8
```

Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c:265]: My mistake!! I deleted it and didn't do a double check. sorry about that...😓😓I have pushed it and checked it again

File: bsp/m5stack_core/include/bsp/m5stack_core.h [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+/**
+ * @brief Check if battery is charging
+ *
+ * @return
+ *      - 0: Not charging
+ *      - 1: Still charging
+ *      - 2: Fully charged
+ *      - 3: Error getting status
+ */
+uint8_t bsp_battery_is_charging(void);
```
Comment (espzav) [inline bsp/m5stack_core/include/bsp/m5stack_core.h:324]: This is not right BSP API.
@tore-espressif Do we want to leave it here or not?

File: bsp/m5stack_core/m5stack_core.c [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+{
+    // Configure backlight GPIO
+    gpio_config_t io_conf = {
+        .pin_bit_mask = (1ULL << BSP_LCD_BACKLIGHT),
+        .mode = GPIO_MODE_OUTPUT,
+        .pull_up_en = GPIO_PULLUP_DISABLE,
+        .pull_down_en = GPIO_PULLDOWN_DISABLE,
+        .intr_type = GPIO_INTR_DISABLE,
+    };
+    return gpio_config(&io_conf);
```
[the remark this answers is not in the document]
Comment (Ocean-lhy) [reply bsp/m5stack_core/m5stack_core.c]: 👌

File: bsp/m5stack_core/include/bsp/m5stack_core.h (hunk 15 above)
Comment (Ocean-lhy) [reply bsp/m5stack_core/include/bsp/m5stack_core.h]: It's a non-standard API indeed, and I've removed it

**Review by espzav — APPROVED** [review]
@Ocean-lhy Thank you for all changes! LGTM

**Pull request merged**